yt_dlp/networking/_helper.py

```diff
@@ -206,3 +207,54 @@ def wrapper(self, *args, **kwargs):
                 e.handler = self
             raise
     return wrapper
+
+
+def _socket_connect(ip_addr, timeout, source_address):
+    af, socktype, proto, canonname, sa = ip_addr
+    sock = socket.socket(af, socktype, proto)
+    try:
+        if timeout is not socket._GLOBAL_DEFAULT_TIMEOUT:
+            sock.settimeout(timeout)
+        if source_address:
+            sock.bind(source_address)
+        sock.connect(sa)
+        return sock
+    except socket.error:
+        sock.close()
+        raise
+
+
+def create_connection(address, timeout=socket._GLOBAL_DEFAULT_TIMEOUT, source_address=None, *, _create_socket_func=_socket_connect):
+    # This is to workaround _create_connection() from socket where it will try all
+    # address data from getaddrinfo() including IPv6. This filters the result from
+    # getaddrinfo() based on the source_address value.
+    # This is based on the cpython socket.create_connection() function.
+    # https://github.com/python/cpython/blob/main/Lib/socket.py#L810
+    host, port = address
+    err = None
+    ip_addrs = addrs = socket.getaddrinfo(host, port, 0, socket.SOCK_STREAM)
+    if source_address is not None:
+        af = socket.AF_INET if '.' in source_address[0] else socket.AF_INET6
+        ip_addrs = [addr for addr in addrs if addr[0] == af]
+        if addrs and not ip_addrs:
+            raise OSError(
+                f"No remote IPv{4 if af == socket.AF_INET else 6} addresses available for connect. "
+                f"Can't use '{source_address[0]}' as source address")
+
+    for ip_addr in ip_addrs:
+        try:
+            sock = _create_socket_func(ip_addr, timeout, source_address)
+            # Break explicitly a reference cycle
```

Comment:
This puzzled me when I first read it. We should declare explicitly what reference we are breaking from (applies both here and down below):
```suggestion
            # Explicitly break __traceback__ reference cycle
```
Cycles will also eventually be collected by the GC, so I'm not sure if we would even need this optimization.

Reply:
Note that this create_connection code is just moved from existing code, which was adapted from CPython. This is present in CPython.

Reply:
Yes, but the [related issue](https://bugs.python.org/issue36820) mentiones this as only an optimization and also tells us this is explicitly about the reference of `__traceback__`.

Maybe link it to that issue? I was at a total loss of what this was suppose to do and had to git blame.